src/Symfony/Bundle/FrameworkBundle/Test/SessionHelperTrait.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Test;
+
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * Provides method to deal with sessions in a stateless container.
+ */
+trait SessionHelperTrait
+{
+    private function callInRequestContext(KernelBrowser $client, callable $callable)
+    {
+        /** @var EventDispatcherInterface $eventDispatcher */
+        $eventDispatcher = static::getContainer()->get(EventDispatcherInterface::class);
+        $return = null;
+        $wrappedCallable = function (RequestEvent $event) use (&$callable, &$return) {
+            try {
+                $return = $callable();
+            } finally {
+                $event->setResponse(new Response(''));
+                $event->stopPropagation();
+            }
+        };
+
+        $eventDispatcher->addListener(KernelEvents::REQUEST, $wrappedCallable);
+        try {
+            $client->request('GET', '/'.uniqid('', true));
```

Comment:
what happens if the routing does not match ? wouldn't this return a 404 *before* calling the callable ?

Reply:
hmm nothing, because the EventListener returns a response AND stops the propagation before the router handles it.